nni/common/concrete_trace_utils/concrete_proxy.py

```diff
@@ -0,0 +1,422 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+import dis
+import logging
+import inspect
+import operator
+
+from typing import List, Optional, Iterable, Any, Set, Union
+
+import torch
+from torch.fx._compatibility import compatibility
+from torch.fx.graph import magic_methods, reflectable_magic_methods
+from torch.fx.node import Node
+from torch.fx.proxy import Proxy
+from torch.overrides import is_tensor_method_or_property
+
+from . import concrete_tracer as et
+from .utils import (
+    _orig_tuple,
+    _orig_list,
+    _orig_type,
+    _orig_isinstance,
+    _orig_getattr,
+    _orig_range,
+    _orig_dict,
+    _orig_len,
+    _orig_index,
+    _orig_bool,
+    _orig_slice,
+    _orig_set,
+    map_recursive,
+)
+
+_logger = logging.getLogger(__name__)
+
+@compatibility(is_backward_compatible=True)
+class ConcreteProxy(Proxy):
+    """
+    `ConcreteProxy` is a wrapped proxy carried the real intermediate value.
+    We can use it to trace a more compatibal model, and pass the branches.
+    """
+
+    # TODO: python bytecode changes a lot in version 3.11. these ops should be updated.
+    jump_opnames = (
+        'JUMP_IF_FALSE_OR_POP',
+        'JUMP_IF_TRUE_OR_POP',
+        'POP_JUMP_IF_FALSE',
+        'POP_JUMP_IF_TRUE',
+        'JUMP_IF_NOT_EXC_MATCH', # occurred in new python vertion, not tested
+    )
+    jump_opcodes = _orig_tuple(dis.opmap[name] for name in jump_opnames if name in dis.opmap)
+    op_compare = dis.opmap['COMPARE_OP']
+    op_extended_arg = dis.opmap['EXTENDED_ARG']
+    op_call = dis.opmap['CALL_FUNCTION']
+    op_call_ex = dis.opmap['CALL_FUNCTION_EX']
+    op_not = dis.opmap['UNARY_NOT']
+    op_unpack_sequence = dis.opmap['UNPACK_SEQUENCE']
+    jump_before_opcodes = (op_compare, op_not)
+
+    # occurred in different python versions
+    op_list_extend = dis.opmap['LIST_EXTEND'] if 'LIST_EXTEND' in dis.opmap else None
+    op_tuple_unpack_call = dis.opmap['BUILD_TUPLE_UNPACK_WITH_CALL'] if 'BUILD_TUPLE_UNPACK_WITH_CALL' in dis.opmap else None
+
+    def __init__(self, node: Node, value: Any, tracer: Optional[et.ConcreteTracer] = None):
+        if tracer is None:
+            # This allows you to create a ConcreteProxy object around a raw Node
+            tracer = et.GraphAppendingConcreteTracer(node.graph)
+        self.tracer = tracer
+        self.value = value
+        self.node = node
+
+    def __repr__(self) -> str:
+        # to detect if in debugging or in code
+        calling_frame_name = inspect.stack()[1][1]
+        if calling_frame_name.endswith('pydevd_exe2.py') or calling_frame_name.endswith('pydevd_safe_repr.py'):
```

Comment:
what are "pydevd_exe2.py" and "pydevd_safe_repr.py"?

Reply:
if the func called in 'debug window', we print the 'proxy' info; others we use the normal value info.
we detect the func caller with 'pydevd_exe2.py' and 'pydevd_safe_repr.py'